compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs

```diff
@@ -1955,6 +1964,410 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                        .iter()
+                        .filter(|field_def| {
+                            let field_type: Ty<'tcx> = self.tcx.type_of(field_def.did);
+
+                            // Only retain fields which mention the blameable generics.
+                            struct_def_generics
+                                .iter()
+                                .any(|param| find_param_def_in_ty_walker(field_type.walk(), param))
+                        })
+                        .collect();
+
+                    if blameable_field_defs.len() != 1 {
+                        // We must have a single blameable field, in order to be able to blame it.
+                        return Err(expr);
+                    }
+
+                    let blameable_field_def: &ty::FieldDef = blameable_field_defs[0];
+                    let blameable_field_index = struct_variant_def
+                        .fields
+                        .iter()
+                        .position(|field_def| field_def == blameable_field_def)
+                        .expect("exists in list");
+
+                    if blameable_field_index < call_ctor_args.len() {
+                        return Ok(&call_ctor_args[blameable_field_index]);
+                    }
+
+                    // We failed to find a matching field in the original struct expression.
+                    Err(expr)
+                }
+                _ => Err(expr),
+            },
+            hir::ExprKind::Tup(elements) => {
+                if let &[single_index] = relevant_ty_args_indices.as_slice() {
+                    if single_index < elements.len() {
+                        return Ok(&elements[single_index]);
+                    }
+                }
+                return Err(expr);
+            }
+            _ => Err(expr),
+        }
+    }
+
+    fn point_at_specific_expr_if_possible_for_predicate_obligation(
+        &self,
+        obligation: &PredicateObligation<'tcx>,
+        expr: &'tcx hir::Expr<'tcx>,
+    ) -> Result<&'tcx hir::Expr<'tcx>, &'tcx hir::Expr<'tcx>> {
+        self.point_at_specific_expr_if_possible_for_obligation_cause_code(
+            obligation.cause.code(),
+            expr,
+        )
+    }
+
+    /**
+     * Recursively searches for the most-specific blamable expression.
+     * Specifically, looks for constructor expressions, e.g. `(false, 5)` constructs a tuple
+     * or `Some(5)` constructs an `Option<i32>` and pairs them with their derived impl predicates.
+     */
+    fn point_at_specific_expr_if_possible(
+        &self,
+        error: &mut traits::FulfillmentError<'tcx>,
+        expr: &'tcx hir::Expr<'tcx>,
+    ) {
+        let expr = match self
+            .point_at_specific_expr_if_possible_for_predicate_obligation(&error.obligation, expr)
+        {
+            Ok(expr) => expr,
+            Err(expr) => expr,
+        };
+
+        // update the error span.
+        error.obligation.cause.span = expr
```

Comment:
Hm, so the span diverging from the actual hir id of the new obligation cause is very suspicious. It'll probably cause incorrect suggestions, since those can look at the *type* of the argument in the obligation cause.